fbink.c

```diff
@@ -5741,13 +5696,65 @@ static __attribute__((cold)) int
 		default:
 			free(font_info->data);
 			free(font_info);
-			WARN("Cannot load font `%s`: requested style (%d) is invalid!", filename, style);
+			WARN("Cannot load font: requested style (%d) is invalid!", style);
 			return ERRCODE(EXIT_FAILURE);
 	}
 
-	ELOG("Font `%s` loaded for style '%s'", filename, font_style_to_string(style));
+	ELOG("Font loaded for style '%s'", font_style_to_string(style));
 	return EXIT_SUCCESS;
 }
+
+// Load OT fonts from a file for fbink_add_ot_font & fbink_add_ot_font_v2
+static __attribute__((cold)) int
+    add_ot_font(const char* filename, FONT_STYLE_T style, FBInkOTFonts* restrict ot_fonts)
+{
+#	ifdef FBINK_FOR_KOBO
+	// NOTE: Bail if we were passed a Kobo system font, as they're obfuscated,
+	//       and some of them risk crashing stbtt because of bogus data...
+	const char blacklist[] = "/usr/local/Trolltech/QtEmbedded-4.6.2-arm/lib/fonts/";
+	if (!strncmp(filename, blacklist, sizeof(blacklist) - 1)) {
+		WARN("Cannot use font `%s`: it's an obfuscated Kobo system font", filename + sizeof(blacklist) - 1);
+		return ERRCODE(EXIT_FAILURE);
+	}
+#	endif
+
+
+	// Open font from given path, and load into buffer
+	FILE* f                      = fopen(filename, "r" STDIO_CLOEXEC);
+	unsigned char* restrict data = NULL;
+	if (!f) {
+		PFWARN("fopen: %m");
+		otInit = false;
+		return ERRCODE(EXIT_FAILURE);
+	} else {
+		const int   fd = fileno(f);
+		struct stat st;
+		if (fstat(fd, &st) == -1) {
+			PFWARN("fstat: %m");
+			fclose(f);
+			otInit = false;
+			return ERRCODE(EXIT_FAILURE);
+		}
+		data = calloc((size_t) st.st_size, sizeof(*data));
+		if (!data) {
+			PFWARN("Error allocating font data buffer: %m");
+			fclose(f);
+			otInit = false;
+			return ERRCODE(EXIT_FAILURE);
+		}
+		if (fread(data, 1U, (size_t) st.st_size, f) < (size_t) st.st_size || ferror(f) != 0) {
+			free(data);
+			fclose(f);
+			otInit = false;
+			WARN("Error reading font file `%s`", filename);
+			return ERRCODE(EXIT_FAILURE);
+		}
+		fclose(f);
+	}
+	int result = add_ot_font_data(data, style, ot_fonts);
+	free(data);
```

Comment:
That's a problem ;).

(The data is still live, stbtt doesn't make a copy, it just keeps a reference to that same pointer).